src/arch/x86/object/ioport.c

```diff
@@ -184,31 +185,13 @@ static exception_t invokeX86PortIn(word_t invLabel, uint16_t port, bool_t call)
     }
 
     if (call) {
-        setRegister(NODE_STATE(ksCurThread), badgeRegister, 0);
-
-        if (n_msgRegisters < 1) {
-            word_t *ipcBuffer;
-            ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
-            if (ipcBuffer != NULL) {
-                ipcBuffer[1] = res;
-                len = 1;
-            } else {
-                len = 0;
-            }
-        } else {
-            setRegister(NODE_STATE(ksCurThread), msgRegisters[0], res);
-            len = 1;
-        }
```

Comment:
That is true, but it'd be a larger verification change and should be done consistently everywhere as in Anna's PR.

Currently we have a lot of different ways kernel replies are generated, and the more there are the higher the chance someone gets something subtly wrong, esp when it's copy/pasted into another context etc.

I would actually be up for taking another look at Anna's PR from then and pulling that into the verification. Probably in smaller steps. The next step from here would be to remove the `isReceiver` parameter from `lookupIPCBuffer` what makes it possible to move the cap rights check up to the place where the IPC buffer cap is set. That might already gain some speed and simplify things a bit. After that, we could look at getting rid of more of the lookup overall. All of those will be easier to do when the conventions are the same as far as possible.

Reply:
Seeing the pattern above here, we could either change lookupIPCBuffer or make an additional version that takes the length of the message (if known statically) and skips the lookup + return NULL when the IPC buffer isn't needed. This would concentrate that logic in one place, which can be architecture dependent, but used generically.  

Not quite sure it'd work everywhere and would optimise out cleanly, but it might be worth trying because that would be a fairly local and systematic change, i.e. once you have figured out what the proof change is for this, you can probably apply it relatively easily everywhere.

Reply:
That would be for a separate PR, though, I think we should merge this one to make sure we have a correct version first.